ci: add secondary MSRV job of 1.75.0

oleonardolima opened the pull request:
partially fixes #1750

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Description

<!-- Describe the purpose of this PR, what's being adding and/or fixed -->

It adds a secondary MSRV version of `1.75.0` for the `bdk_electrum` crate, excluding it from the `1.63.0` MSRV dependency pinning and CI step.

As it's the new CI will run without `bdk_electrum` for `1.63.0` MSRV step, will run another step for `1.75.0` MSRV and the latest stable.

Currently,, the only affected crate is `electrum-client`, therefore `bdk-electrum`, as it relies directly on `rustls` which migrated to `1.71.0` MSRV. The `esplora-client` relies on `rustls` as a dependency of `minreq` or `reqwest`, but those crates didn't upgraded it's `rustls` versions yet.

In a further PR it should also bump the `electrum-client` crate to it's latest version, which relies on `1.75.0` MSRV, see https://github.com/bitcoindevkit/rust-electrum-client/pull/159.

### Notes to the reviewers

<!-- In this section you can include notes directed to the reviewers, like explaining why some parts
of the PR were done in a specific way -->

It's open for discussion if this approach is the expected one, or another one would be better.

### Changelog notice

<!-- Notice the release manager should include in the release tag message changelog -->
<!-- See https://keepachangelog.com/en/1.0.0/ for examples -->

- Add a secondary MSRV of `1.75.0` for `bdk_electrum`.

### Checklists

#### All Submissions:

* [x] I've signed all my commits
* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
* [x] I ran `cargo fmt` and `cargo clippy` before committing

#### New Features:

* [ ] I've added tests for the new feature
* [ ] I've added docs for the new feature

#### Bugfixes:

* [ ] This pull request breaks the existing API
* [ ] I've added tests to reproduce the issue which are now passing
* [ ] I'm linking the issue being fixed by this PR

oleonardolima committed:
ci: add secondary MSRV job of 1.75.0

oleonardolima committed:
chore(cleanup): remove rustls from pinned deps

ValuedMammal committed:
docs: update `README.md` with MSRV policy

- bdk_electrum: add MSRV section to README.md

ValuedMammal committed:
ci: add `pin-msrv.sh`

- add ci/pin-msrv.sh, which pins the dependencies for 1.63.0 MSRV

ValuedMammal commented:
I sort of like the idea of separating the "check" and "test" jobs. For example this CI run https://github.com/ValuedMammal/bdk/actions/runs/12872751304

oleonardolima commented:
> I sort of like the idea of separating the "check" and "test" jobs. For example this CI run https://github.com/ValuedMammal/bdk/actions/runs/12872751304

That's a good idea, I'll cherry-pick the commit.

oleonardolima commented:
> > I sort of like the idea of separating the "check" and "test" jobs. For example this CI run https://github.com/ValuedMammal/bdk/actions/runs/12872751304
> 
> That's a good idea, I'll cherry-pick the commit.

However, @ValuedMammal any reason to not have tests run with `--no-default-features` too?

ValuedMammal commented:
> > > I sort of like the idea of separating the "check" and "test" jobs. For example this CI run https://github.com/ValuedMammal/bdk/actions/runs/12872751304
> > 
> > 
> > That's a good idea, I'll cherry-pick the commit.
> 
> However, @ValuedMammal any reason to not have tests run with `--no-default-features` too?

I thought the same thing and I don't know. It didn't seem like that would pick up any tests that weren't already included in `--all-features`, but in any case I'm in favor of running all possible tests.

oleonardolima commented:
> > > > I sort of like the idea of separating the "check" and "test" jobs. For example this CI run https://github.com/ValuedMammal/bdk/actions/runs/12872751304
> > > 
> > > 
> > > That's a good idea, I'll cherry-pick the commit.
> > 
> > 
> > However, @ValuedMammal any reason to not have tests run with `--no-default-features` too?
> 
> I thought the same thing and I don't know. It didn't seem like that would pick up any tests that weren't already included in `--all-features`, but in any case I'm in favor of running all possible tests.

I cherry-picked and add the features to matrix, but it now fails because not all crates has `miniscript/no-std`, I'll need to sort that out.

oleonardolima commented:
Alright, I kept it all under a single build-test job, using the `./ci/pin-msrv.sh` script and referring to it on README.md.

It should be ready for review.

notmandatory commented on README.md:
````diff
@@ -63,25 +63,17 @@ Fully working examples of how to use these components are in `/example-crates`:
-cargo update -p unicode-width --precise "0.1.13"
-cargo update -p rustls@0.23.21 --precise "0.23.19"
-```
+The BDK library maintains a MSRV of 1.63.0. This includes the following crates 
````
```suggestion
The BDK library maintains a MSRV of 1.63.0. This includes the following crates:
```

notmandatory commented on README.md:
```diff
@@ -63,25 +63,17 @@ Fully working examples of how to use these components are in `/example-crates`:
+
+The MSRV of `bdk_electrum` is 1.75.0.
+
+To build with the MSRV of 1.63.0 you will need to pin dependencies by running the [`pin-msrv.sh`](./ci/pin-msrv.sh) script
```
```suggestion
To build with the MSRV of 1.63.0 you will need to pin dependencies by running the [`pin-msrv.sh`](./ci/pin-msrv.sh) script.
```

notmandatory commented on ci/pin-msrv.sh:
```diff
@@ -0,0 +1,24 @@
+# To pin deps, switch toolchain to MSRV and execute the below updates
+
+# cargo clean
+# rustup default 1.63.0
```
I had to use this command:
```suggestion
# rustup override set 1.63.0
```

oleonardolima replied:
It depends on whether the user is using `overrides` for the directory or not 🤔. I guess it's best to recommend using a command that has only side-effects within the directory, will update.

notmandatory reviewed (commented):
Looks good, I found a couple nits to fix and it needs to be rebased now that #1801 is merged.

oleonardolima committed:
chore: fix punctuation and use `rustup override set` instead

- fix some nits and punctuation typos.
- recommend usage of `rustup override set` instead of `rustup default` command.

Co-authored-by: Steve Myers <github@notmandatory.org>

notmandatory reviewed (approved):
ACK c8b55da39913ddaf11d26f09d92034aef4f550e0

notmandatory commented:
I updated Github required checks to match changes in this PR.
